### PR TITLE
Fix metrics context circular reference

### LINKED ISSUE #1534: metrics memory leak because of circular reference
There is circular reference between `meter` and `context`:
https://github.com/open-telemetry/opentelemetry-cpp/runs/7600470707?check_suite_focus=true#step:5:806
This causes a memory leak.

### PR DESCRIPTION
**Pull request by esigo**
Fixes #1534 (issue)

## Changes
removed the cycles by using a `wrak_ptr` to the context in `meter`. Also `meter_collector` can't have `weak_ptr` to the context so I used raw pointer.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by lalitb** [comment]
@esigo  - I have gone through the circular reference between `meter_context <-> meter`, and `meter_context <-> metric_collector`, and I don't think it would be easy to avoid them. These changes look good in that scenario. We can have a separate issue to track the redesign requirement for metrics v1.0 / GA.

File: sdk/include/opentelemetry/sdk/metrics/meter.h [hunk 1]
```diff
@@ -135,7 +135,8 @@ class Meter final : public opentelemetry::metrics::Meter
                                                    void *),
                                   void *state = nullptr)
   {
-    auto view_registry = meter_context_->GetViewRegistry();
+    auto ctx           = meter_context_.lock();
```
[the remark this answers is not in the document]
Comment (esigo) [reply sdk/include/opentelemetry/sdk/metrics/meter.h]: makes sense. Shall we log something too?

File: sdk/src/metrics/meter.cc [hunk 2]
```diff
@@ -201,7 +201,8 @@ const sdk::instrumentationscope::InstrumentationScope *Meter::GetInstrumentation
 std::unique_ptr<WritableMetricStorage> Meter::RegisterMetricStorage(
     InstrumentDescriptor &instrument_descriptor)
 {
-  auto view_registry = meter_context_->GetViewRegistry();
+  auto ctx           = meter_context_.lock();
```
[the remark this answers is not in the document]
Comment (lalitb) [reply sdk/src/metrics/meter.cc:207]: Yes, I think we should do that.

File: sdk/include/opentelemetry/sdk/metrics/meter.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/metrics/meter.h]: yes, I think we should log here if meter context is no more valid.

[the remark this answers is not in the document]
Comment (esigo) [reply sdk/include/opentelemetry/sdk/metrics/meter.h]: done

File: sdk/src/metrics/meter.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (esigo) [reply sdk/src/metrics/meter.cc]: done

**Review by lalitb — APPROVED** [review]

**Pull request merged**